modules/core/04-channel/types/upgrade.go

```diff
@@ -77,12 +72,55 @@ func (ut Timeout) IsValid() bool {
 	return !ut.Height.IsZero() || ut.Timestamp != 0
 }
 
-// NewErrorReceipt returns an error receipt with the code from the provided error type stripped
-// out to ensure changes of the error message don't cause state machine breaking changes.
-func NewErrorReceipt(upgradeSequence uint64, err error) ErrorReceipt {
-	_, code, _ := errorsmod.ABCIInfo(err, false) // discard non-determinstic codespace and log values
+// UpgradeError defines an error that occurs during an upgrade.
+type UpgradeError struct {
+	// err is the underlying error that caused the upgrade to fail.
+	// this error should not be written to state.
+	err error
+	// sequence is the upgrade sequence number of the upgrade that failed.
+	sequence uint64
+}
+
+// NewUpgradeError returns a new UpgradeError instance.
+func NewUpgradeError(upgradeSequence uint64, err error) UpgradeError {
+	return UpgradeError{
+		err:      err,
+		sequence: upgradeSequence,
+	}
+}
+
+// Error implements the error interface, returning the underlying error which caused the upgrade to fail.
+func (u UpgradeError) Error() string {
+	return u.err.Error()
+}
+
+
+// Is returns true if the underlying error is of the given err type.
+func (u UpgradeError) Is(err error) bool {
+	return errors.Is(u.err, err)
+}
+
+// Unwrap returns the base error that caused the upgrade to fail.
+func (u UpgradeError) Unwrap() error {
+	for {
+		if err := errors.Unwrap(u.err); err != nil {
+			u.err = err
+		} else {
+			return u.err
+		}
+	}
+}
```

Comment:
I guess this would help in `Is` since it would unwrap an upgrade error if it is provided as the target?

Reply:
so it looks like an issue we were running into is that if you are just wrapping an sdk error, eg. `ErrInvalidChannel` you don't need to worry about `Unwrap`. The following test code works without implementing `Unwrap` in the `UpgradeError` type

```golang
func (suite *TypesTestSuite) TestGetErrorReceipt() {
	upgradeErr := channeltypes.NewUpgradeError(1, channeltypes.ErrInvalidChannel)
	wrappedUpgradeErr := channeltypes.NewUpgradeError(1, errorsmod.Wrap(channeltypes.ErrInvalidChannel, "testing"))

	suite.Require().Equal(upgradeErr.GetErrorReceipt().Message, wrappedUpgradeErr.GetErrorReceipt().Message)
}
```

However, if you want to be able to wrap the `UpgradeError` itself (which I think there is no harm providing the option) the `Unwrap` method needs to be implemented.

For example, if the `Unwrap` method is removed from `UpgradeError`, the following test fails.

```golang
func (suite *TypesTestSuite) TestGetErrorReceipt() {
	ue := channeltypes.NewUpgradeError(1, channeltypes.ErrInvalidChannel)

	wrappedErr := errorsmod.Wrap(ue, "wrapped upgrade error")
	suite.Require().True(errorsmod.IsOf(wrappedErr, channeltypes.ErrInvalidChannel))

	ue2 := channeltypes.NewUpgradeError(1, wrappedErr)

	suite.Require().Equal(ue2.GetErrorReceipt().Message, ue.GetErrorReceipt().Message)
}
```

with the error

```
-ABCI code: 1: restored channel to pre-upgrade state
+ABCI code: 4: restored channel to pre-upgrade state
```

I we should be able to wrap the `UpgradeError` type, so we don't need to worry about special casing it in any way.

WDYT?

Reply:
Oh yea! It's because `UpgradeError` doesn't implement the `Cause` interface function. See https://github.com/cosmos/ibc-go/pull/3700#discussion_r1213170980